emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
```

Comment:
But from the three backends only Spark's Dataset has native support for that, for the others, you'll have to implement it with `coGroup` similar to what I suggested above.

Also

```
(this minus that).isEmpty && (that minus this).isEmpty
```

will require two co-groups, whereas the variant I sketched can work with one.

My other concern is that minus makes reasoning about logical rewrites a bit harder, and it is not needed for a lot of use cases we can make our life simpler by simply not including it (at least until we see a strong need for the opposite).